### PR TITLE
3.0 Prioritizing cascade delete to assoc. for which cascadeCallbacks is true

### PR DESCRIPTION
**Pull request by PGBI**
To fully understand this pull request, here is the bug i'm trying to fix:
## Setup

Let's say that:
- Users belongsToMany Groups through GroupMemberships
- GroupMemberships table contains a field named `leading` set to true or false whether its user is the leader of its group or not. 
- Users table contains a field named `leading_groups_count`. 

Tables are setup like this:

```
// GroupMemberships.php
$this->belongsTo('Users');
$this->addBehavior('CounterCache', ['Users' => [
    'leading_groups_count' => [
        'conditions' => [
            'leading' => true,
        ]
    ]
]);

// GroupsTable.php
$this->belongsToMany('Users', ['through' => 'GroupMemberships']));
$this->hasMany('GroupMemberships', [
    'dependent' => true,
    'cascadeCallbacks' => true
]);
```
## The problem

When a group is deleted, I'm expecting the related GroupMemberships rows to be deleted one by one (and not through deleteAll()) as I set cascadeCallbacks to true.
But it's not happening this way. GroupMemberships rows are being bulk deleted. As a result, the counterCache behavior is not called, and the `leading_groups_count` field in Users table is not updated.
## Solution proposed

I found that if I defined the HasMany association before the belongsToMany one in GroupsTable, the problem was solved. However, I don't like the fact that the app behavior depends on the order where associations are being defined (it doesn't look right).

So what I did was modifying AssociationCollection::cascadeDelete() to make sure that associations for which cascadeCallbacks is true are handled first.

### CODE REVIEW

**Comment by lorenzo** [comment]
Can you add a test for this?

File: src/ORM/AssociationCollection.php [hunk 1]
```diff
@@ -259,7 +259,14 @@ protected function _save($association, $entity, $nested, $options)
     public function cascadeDelete(Entity $entity, array $options)
     {
         foreach ($this->_items as $assoc) {
-            $assoc->cascadeDelete($entity, $options);
+            if ($assoc->cascadeCallbacks()) {
+                $assoc->cascadeDelete($entity, $options);
+            }
+        }
+        foreach ($this->_items as $assoc) {
```
Comment (markstory) [inline src/ORM/AssociationCollection.php]: Looping twice over the full set of associations seems less than ideal. Could we loop and filter, so the second loop is shorter?

**Comment by PGBI** [comment]
@lorenzo What kind of test would you see here?
- just checking that associations with cascadeCallbacks to true are called first?
- or testing the scenario I described (with Users belongsToMany Groups through GroupsMemberships with a counterCache on GroupsMemberships)?

**Comment by PGBI** [comment]
@markstory I've done the loop and filter as suggested.
@lorenzo I've added a test for this scenario. Sorry I was not very inspired when choosing a name for the test function...

**Comment by markstory** [comment]
Thanks :+1:

File: tests/Fixture/MembersFixture.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * CakePHP(tm) Tests <http://book.cakephp.org/2.0/en/development/testing.html>
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://book.cakephp.org/2.0/en/development/testing.html CakePHP(tm) Tests
+ * @since         1.2.0
```
Comment (ADmad) [inline tests/Fixture/MembersFixture.php]: Copy paste fail :)

Comment (markstory) [inline tests/Fixture/MembersFixture.php]: I'm pretty sure I fixed those in my fixup commit.

**Pull request merged**